File: kits/csharp/simple/Bot/Lux/Agent.cs
using System;
using System.IO;

namespace Bot.Lux
{
	public class Agent
	{
		public GameState GameState = new GameState();
		public TextReader Reader;

		/**
         * Constructor for a new agent User should edit this according to their `Design`
         */
		public Agent()
		{
			Reader = Console.In;
		}

		/**
         * Initialize Agent for the `Match` User should edit this according to their
         * `Design`
         */
		public void Initialize()
		{
			// get agent ID
			GameState.Id = int.Parse(Reader.ReadLine() ?? throw new InvalidOperationException());
			var mapInfo = Console.ReadLine();
			if (mapInfo == null) return;
			var mapInfoSplit = mapInfo.Split(" ");
			var mapWidth = int.Parse(mapInfoSplit[0]);
			var mapHeight = int.Parse(mapInfoSplit[1]);
			GameState.Map = new GameMap(mapWidth, mapHeight);
		}

		/**
         * Updates agent's own known state of `Match` User should edit this according to
         * their `Design`.
         */
		public void Update()
		{
			// wait for the engine to send any updates
			GameState.Map = new GameMap(GameState.Map.Width, GameState.Map.Height);
			GameState.Turn += 1;
			GameState.Players[0].Cities.Clear();
			GameState.Players[0].Units.Clear();
			GameState.Players[1].Cities.Clear();
			GameState.Players[1].Units.Clear();
			while (true)
			{
				var updateInfo = Reader.ReadLine();
				if (updateInfo == IOConstants.DONE)
				{
					break;
				}
				var updates = updateInfo.Split(" ");
				var inputIdentifier = updates[0];
				if (inputIdentifier == IOConstants.RESEARCH_POINTS)
				{
					var team = int.Parse(updates[1]);
					GameState.Players[team].ResearchPoints = int.Parse(updates[2]);
				}
				else if (inputIdentifier == IOConstants.RESOURCES)
				{
					var type = updates[1];
					var x = int.Parse(updates[2]);
					var y = int.Parse(updates[3]);
					var amt = (int)double.Parse(updates[4]);
					GameState.Map.SetResource(type, x, y, amt);
				}
				else if (inputIdentifier == IOConstants.UNITS)
				{
					var i = 1;
					var unitType = int.Parse(updates[i++]);
					var team = int.Parse(updates[i++]);
					var unitId = updates[i++];
					var x = int.Parse(updates[i++]);
					var y = int.Parse(updates[i++]);
					var cooldown = double.Parse(updates[i++]);
					var wood = int.Parse(updates[i++]);
					var coal = int.Parse(updates[i++]);
					var uranium = int.Parse(updates[i++]);
					var unit = new Unit(team, unitType, unitId, x, y, cooldown, wood, coal, uranium);
					GameState.Players[team].Units.Add(unit);
				}
				else if (inputIdentifier == IOConstants.CITY)
				{
					var i = 1;
					var team = int.Parse(updates[i++]);
					var cityId = updates[i++];
					var fuel = double.Parse(updates[i++]);
					var lightUpkeep = double.Parse(updates[i++]);
					GameState.Players[team].Cities.Add(cityId, new City(team, cityId, fuel, lightUpkeep));
				}
				else if (inputIdentifier == IOConstants.CITY_TILES)
				{
					var i = 1;
					var team = int.Parse(updates[i++]);
					var cityId = updates[i++];
					var x = int.Parse(updates[i++]);
					var y = int.Parse(updates[i++]);
					var cooldown = double.Parse(updates[i++]);
					var city = GameState.Players[team].Cities[cityId];
					var cityTile = city.AddCityTile(x, y, cooldown);
					GameState.Map.GetCell(x, y).CityTile = cityTile;
					GameState.Players[team].CityTileCount += 1;
				}
				else if (inputIdentifier == IOConstants.ROADS)
				{
					var i = 1;
					var x = int.Parse(updates[i++]);
					var y = int.Parse(updates[i++]);
					var road = double.Parse(updates[i++]);
					var cell = GameState.Map.GetCell(x, y);
					cell.Road = road;
				}
			}
		}

		/**
         * End a turn
         */
		public void EndTurn()
		{
			Console.Out.Write("D_FINISH");
			Console.Out.Flush();
		}
	}
}

File: kits/fsharp/simple/Bot/Lux/Agent.fs
module Lux.Agent

open System
open GameObjects
open System.Collections.Generic
open System.IO

let private updateUnfold (gameState:Game) (input: TextReader) : option<Game * Game> =
    let updateInfo = input.ReadLine()
    if updateInfo = Constants.INPUT_CONSTANTS.DONE then
        None
    else
        let updates = updateInfo.Split(" ");
        let inputIdentifier : string = updates.[0];
        match inputIdentifier with
            | s when s = Constants.INPUT_CONSTANTS.RESEARCH_POINTS ->
                let team: int = Int32.Parse(updates.[1]);
                let newPoints = Int32.Parse(updates.[2])
                gameState.Players.[team].research_points <- newPoints
                Some(gameState, gameState)
            | s when s = Constants.INPUT_CONSTANTS.RESOURCES ->
                let r_type = updates.[1];
                let x = Int32.Parse(updates.[2]);
                let y = Int32.Parse(updates.[3]);
                let amt = (int)(Double.Parse(updates.[4]))
                gameState.GameMap._setResource(r_type, x, y, amt)
                Some(gameState, gameState)
            | s when s = Constants.INPUT_CONSTANTS.UNITS ->
                let unittype = Int32.Parse(updates.[1]);
                let team = Int32.Parse(updates.[2]);
                let unitid = updates.[3];
                let x = Int32.Parse(updates.[4]);
                let y = Int32.Parse(updates.[5]);
                let cooldown = Double.Parse(updates.[6]);
                let wood = Int32.Parse(updates.[7]);
                let coal = Int32.Parse(updates.[8]);
                let uranium = Int32.Parse(updates.[9]);
                let unit = new Unit(team, Constants.ParseUnitType unittype, unitid, x, y, cooldown, wood, coal, uranium);
                let newGameState = 
                    match gameState.Players with
                        | players ->
                            players.[team].Units <- unit::players.[team].Units
                            { gameState with 
                                Players = players
                            }
                        | _ -> failwith "More than two players present, do ghosts exits?"
                Some(newGameState, newGameState)
            | s when s = Constants.INPUT_CONSTANTS.CITY ->
                let team = Int32.Parse(updates.[1]);
                let cityid = updates.[2];
                let fuel = Double.Parse(updates.[3]);
                let lightUpkeep = Double.Parse(updates.[4]);
                gameState.Players.[team].Cities.Add(cityid, new City(team, cityid, fuel, lightUpkeep));
                Some(gameState, gameState)
            | s when s = Constants.INPUT_CONSTANTS.CITY_TILES ->
                let team = Int32.Parse(updates.[1]);
                let cityid = updates.[2];
                let x = Int32.Parse(updates.[3]);
                let y = Int32.Parse(updates.[4]);
                let cooldown = Double.Parse(updates.[5]);
                let city = gameState.Players.[team].Cities.Item cityid
                let citytile = city._add_city_tile(x, y, cooldown);
                gameState.GameMap.GetCell(x, y).Citytile <- citytile;
                gameState.Players.[team].city_tile_count <- gameState.Players.[team].city_tile_count + 1;
                Some(gameState, gameState)
            | s when s = Constants.INPUT_CONSTANTS.ROADS ->
                let x = Int32.Parse(updates.[1]);
                let y = Int32.Parse(updates.[2]);
                let road = Double.Parse(updates.[3]);
                let cell = gameState.GameMap.GetCell(x, y);
                cell.Road <- road;
                Some(gameState, gameState)
            | _ -> failwith "Unknown input string"

type IGameStateIO =
    abstract member update : unit -> Game
    abstract member endTurn: unit -> unit

type GameStateIO =
    {
        gameState : Game
    }
    static member initialize(input: TextReader) : Game =
        let id = Int32.Parse(input.ReadLine())
        let mapInfo: string = input.ReadLine()
        let mapInfoSplit: string[] = mapInfo.Split(" ")
        let mapWidth = Int32.Parse(mapInfoSplit.[0])
        let mapHeight = Int32.Parse(mapInfoSplit.[1])
        let map = new GameMap(mapWidth, mapHeight)
        let gameState : Game = 
            {
                Id = id 
                GameMap = map;
                Turn = 0;
                Players = [Player(0); Player(1)]
                }
        gameState

    member self.update(input: TextReader) =
        let newPlayer0 = 
            Player(0, [], new Dictionary<string, City>(), self.gameState.Players.[0].city_tile_count)
        let newPlayer1 = 
            Player(1, [], new Dictionary<string, City>(), self.gameState.Players.[1].city_tile_count)
        let newGameState = 
            { self.gameState with
                Turn = self.gameState.Turn + 1;
                Players = [newPlayer0; newPlayer1]
                GameMap = GameMap(self.gameState.GameMap.height, self.gameState.GameMap.width)
                }
            |> Seq.unfold (fun state ->
                updateUnfold state input
                )
            |> Seq.last
        { self with gameState = newGameState }

    /// Ends turn
    member self.endTurn() =
        Console.WriteLine("D_FINISH")
